src/drivers/imx/edma.c

```diff
@@ -3,115 +3,520 @@
 // Copyright 2019 NXP
 //
 // Author: Daniel Baluta <daniel.baluta@nxp.com>
+// Author: Paul Olaru <paul.olaru@nxp.com>
 
+#include <sof/audio/component.h>
 #include <sof/drivers/edma.h>
+#include <sof/drivers/timer.h>
+#include <sof/lib/alloc.h>
 #include <sof/lib/dma.h>
 #include <sof/lib/io.h>
+#include <sof/platform.h>
 #include <stdint.h>
 
-static inline void edma_write(struct dma *dma, uint32_t reg, uint32_t value)
+static int edma_compute_attr(int src_width, int dest_width)
 {
-	io_reg_write(dma_base(dma) + reg, value);
-}
+	int result = 0;
 
-static inline uint32_t edma_read(struct dma *dma, uint32_t reg)
-{
-	return io_reg_read(dma_base(dma) + reg);
-}
+	switch (src_width) {
+	case 1:
+		result = EDMA_TCD_ATTR_SSIZE_8BIT;
+		break;
+	case 2:
+		result = EDMA_TCD_ATTR_SSIZE_16BIT;
+		break;
+	case 4:
+		result = EDMA_TCD_ATTR_SSIZE_32BIT;
+		break;
+	case 8:
+		result = EDMA_TCD_ATTR_SSIZE_64BIT;
+		break;
+	case 16:
+		result = EDMA_TCD_ATTR_SSIZE_16BYTE;
+		break;
+	case 32:
+		result = EDMA_TCD_ATTR_SSIZE_32BYTE;
+		break;
+	case 64:
+		result = EDMA_TCD_ATTR_SSIZE_64BYTE;
+		break;
+	default:
+		return -EINVAL;
+	}
 
-static inline void edma_write16(struct dma *dma, uint32_t reg, uint16_t value)
-{
-	io_reg_write16(dma_base(dma) + reg, value);
-}
-
-static inline uint16_t edma_read16(struct dma *dma, uint32_t reg)
-{
-	return io_reg_read16(dma_base(dma) + reg);
-}
+	switch (dest_width) {
+	case 1:
+		result |= EDMA_TCD_ATTR_DSIZE_8BIT;
+		break;
+	case 2:
+		result |= EDMA_TCD_ATTR_DSIZE_16BIT;
+		break;
+	case 4:
+		result |= EDMA_TCD_ATTR_DSIZE_32BIT;
+		break;
+	case 8:
+		result |= EDMA_TCD_ATTR_DSIZE_64BIT;
+		break;
+	case 16:
+		result |= EDMA_TCD_ATTR_DSIZE_16BYTE;
+		break;
+	case 32:
+		result |= EDMA_TCD_ATTR_DSIZE_32BYTE;
+		break;
+	case 64:
+		result |= EDMA_TCD_ATTR_DSIZE_64BYTE;
+		break;
+	default:
+		return -EINVAL;
+	}
 
-static inline void edma_update_bits(struct dma *dma, uint32_t reg,
-				    uint32_t mask, uint32_t value)
-{
-	io_reg_update_bits(dma_base(dma) + reg, mask, value);
+	return result;
 }
 
 /* acquire the specific DMA channel */
 static struct dma_chan_data *edma_channel_get(struct dma *dma,
 					      unsigned int req_chan)
 {
-	return NULL;
+	struct dma_chan_data *channel;
+
+	tracev_edma("EDMA: channel_get(%d)", req_chan);
+	if (req_chan >= dma->plat_data.channels) {
+		trace_edma_error("EDMA: Channel %d out of range", req_chan);
+		return NULL;
+	}
+
+	channel = &dma->chan[req_chan];
+	if (channel->status != COMP_STATE_INIT) {
+		trace_edma_error("EDMA: Cannot reuse channel %d", req_chan);
+		return NULL;
+	}
+
+	channel->status = COMP_STATE_READY;
+	return channel;
 }
 
 /* channel must not be running when this is called */
 static void edma_channel_put(struct dma_chan_data *channel)
 {
+	/* Assuming channel is stopped, we thus don't need hardware to
+	 * do anything right now
+	 */
+	tracev_edma("EDMA: channel_put(%d)", channel->index);
```

Comment:
There is a check for this inside edma_channel_get. So, all the calls for edma_channel_put will have a valid channel.

Reply:
Yes, for sure, but that's precisely the purpose of asserts to checks to check our SW invariants without trusting to much on the outside world. 
Here, EDMA returned a god pointer to SOF, but there is no reason to suppose that this pointer won't ever be lost or corrupted by some bug.

Reply:
Well there isn't actually _any_ way to verify that this pointer points to some channel, besides manually comparing with the address of all 32 (or 16). And are you really going to do such a laborious check on a function that may be called relatively often? I don't think so.

Reply:
I just mean it's good practice, especially for rare management actions where the cost of the check is negligible, and in particular at the boundaries of the module.

Reply:
If the pointer was wrong we would have had issues getting to this callback itself (a valid pointer to one of the 32 channels is required to even do the dispatch. If the pointer was wrong, then the dma_channel_put function would use an invalid dmac pointer itself and then wouldn't just find a stray pointer to lead to this function.

If the channel pointer was wrong then we wouldn't have reached this point.

So, all callbacks which take a channel* parameter can safely assume that it's valid.

All the required checks of this kind are done. In dma_channel_get we do check the index (and Daniel just added the proper locking), in everything else if the pointer is nonnull (though that is already checked).

Finally, if I quote you:

> Here, EDMA returned a god pointer to SOF, but there is no reason to suppose that this pointer won't ever be lost or corrupted by some bug.

Well the loss or corruption of the pointer is the rest of the firmware's responsibility, not this driver's. If there indeed is a bug from that we would not reach here anyway but either crash or do even weirder stuff than simply entering the correct callback function normally. If there is a bug we fix it where it is, not do *false* defense against it. Because trying to even read channel->index with an invalid pointer is just not gonna be correct (who says that we will be guaranteed to even read an out of range value? We could read a zero and then assume the channel is correct)

Reply:
So no, this check isn't required and cannot even be done reliably. Where it can be done reliably and quickly it _is_ done.

Reply:
But I also see that in the INTEL's driver code (hda-dma.c for example) they also don't do such checks.
Actually it's their cup of tea, for me that's ok to not have such asserts as long as I don't debug on their platforms...  :-)

Reply:
My issue is that you CANNOT validate pointers. So no reliable check can be done.

Will resolve this conversation.

Do tell, what `assert` you would have inserted, and I would have told you why it wouldn't have worked.

Reply:
Lets take this offline, I'm pretty sure all of us have good points to back up our ideas. We could add validation in later patches, with proper explanations.

This is really not a blocker for now.